src/coordinator/amoc_coordinator.erl

```diff
@@ -0,0 +1,170 @@
+start(Name, CoordinationPlan, Timeout) when ?IS_TIMEOUT(Timeout) ->
+    Plan = normalize_coordination_plan(CoordinationPlan),
+    OrderedPlan = order_plan(Plan),
+    case amoc_coordinator_sup:start_coordinator(Name, OrderedPlan, Timeout) of
+        {ok, _} ->
+            amoc_telemetry:execute([coordinator, start], #{count => 1}, #{name => Name}),
+            ok;
+        {error, _} -> error
+    end.
+
+%% @doc Stops a coordinator.
+-spec stop(name()) -> ok.
+stop(Name) ->
+    amoc_coordinator_sup:stop_coordinator(Name),
+    amoc_telemetry:execute([coordinator, stop], #{count => 1}, #{name => Name}).
+
+%% @see add/3
+-spec add(name(), any()) -> ok.
+add(Name, Data) ->
+    add(Name, self(), Data).
+
+%% @doc Adds the current process data. Usually called in the `start/2' callback of an amoc scenario.
+-spec add(name(), pid(), any()) -> ok.
+add(Name, Pid, Data) ->
+    notify(Name, {coordinate, {Pid, Data}}).
+
+%% @doc Resets a coordinator, that is, calls all coordination actions with `reset' as the coordination data.
+-spec reset(name()) -> ok.
+reset(Name) ->
+    notify(Name, reset_coordinator).
+
+-spec notify(name(), coordinator_timeout | reset_coordinator | {coordinate, {pid(), term()}}) -> ok.
+notify(Name, coordinator_timeout) when is_atom(Name) ->
+    do_notify(Name, coordinator_timeout);
+notify(Name, reset_coordinator) when is_atom(Name) ->
+    do_notify(Name, reset_coordinator);
+notify(Name, {coordinate, _} = Event) when is_atom(Name) ->
+    do_notify(Name, Event).
+
+do_notify(Name, Event) ->
+    raise_telemetry_event(Name, Event),
+    case amoc_coordinator_sup:get_workers(Name) of
+        {ok, TimeoutWorker, Workers} ->
+            amoc_coordinator_timeout:notify(TimeoutWorker, Event),
+            [ notify_worker(Worker, Event) || Worker <- Workers ],
+            ok;
+        {error, not_found} ->
+            ok
+    end.
+
+%%%===================================================================
+%%% Internal functions
+%%%===================================================================
+notify_worker(WorkerPid, coordinator_timeout) -> %% synchronous notification
+    amoc_coordinator_worker:timeout(WorkerPid);
+notify_worker(WorkerPid, reset_coordinator) -> %% synchronous notification
+    amoc_coordinator_worker:reset(WorkerPid);
+notify_worker(WorkerPid, {coordinate, {Pid, Data}}) when is_pid(Pid) -> %% async notification
+    amoc_coordinator_worker:add(WorkerPid, {Pid, Data}).
+
+raise_telemetry_event(Name, Event) ->
+    TelemetryEvent = case Event of
+                         {coordinate, _} -> add;
+                         reset_coordinator -> reset;
+                         coordinator_timeout -> timeout
+                     end,
+    amoc_telemetry:execute([coordinator, TelemetryEvent], #{count => 1}, #{name => Name}).
+
+-spec order_plan([normalized_coordination_item()]) -> [normalized_coordination_item()].
+order_plan(Items) ->
```

Comment:
please add a test for this function.
I must say that the version with `lists:partition/2` was more verbose.

+ The main idea of this function is that it defines an order of execution, the rest code must ensure that the order is kept and not broken. all other comments are irrelevant here. internal implementation can change, but execution order must remain as defined here. also, with the latest changes we no longer rely on the childrens' order.

Reply:
What we were doing before was respective the relative order given by the user between non-`all` elements and `all` elements, and simply moving the `all` items to the end. It is the same effect now, so this new code sets an equivalent order that it was already.

Reply:
yeah, I know that it's the same, but lists:partition/2 version was easier to read.

Reply:
All right, reverted then 👌🏽

Reply:
please update the comment for order_plan/1 function:

```
The main idea of this function is that it defines an order of execution, the rest code must ensure that the order is kept and not broken. all other comments are irrelevant here. internal implementation can change, but execution order must remain as defined here. also, with the latest changes we no longer rely on the children's order.
```

I would suggest a comment like this:
```
This function defines the execution order of the events that must be processed synchronously (on reset/timeout/stop). we need to ensure that 'all' items are executed after 'non-all'.
Note that '{coordinate, _}' events are exceptional and their processing is done asynchronously.
The order of All plan items must be preserved as in the original plan, the same applies to NonAll items.
```